File: data-structures/hash-tables/c-sharp/hash_function_string.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{

    class HashTables
    {
        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            // Generate a hash value from an integer hash key
            string key = "A5RD";
            int slots = 97;
            int value = HashString(key, slots);
            Console.WriteLine($"Hash value for {key} is {value}");
        }

        // Produce a hash value from a string
        public static int HashString(string hashKey, int numberOfSlots)
        {
            // Initialise total
            int total = 0;

            // Repeat for every character in the hash key
            for (int i = 0; i < hashKey.Length; i++) {
                // Add the character's ASCII value to the total
                int asciiCode = (int)hashKey[i];
                total = total + asciiCode;
            }

            // Generate the hash value using the modulo operator
            int hashValue = total % numberOfSlots;

            // Return the hash value
            return hashValue;
        }


    }
}

File: data-structures/hash-tables/python/hash_function_string.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


def hash_string(hash_key, number_of_slots):
    """Produce a hash value from a string using the ASCII values"""

    # Initialise total
    total = 0

    # Repeat for every character in the hash key
    for i in range(len(hash_key)):
        # Add the character's ASCII value to the total
        ascii_code = ord(hash_key[i])
        total = total + ascii_code
        
    # Generate the hash value using the modulo operator
    hash_value = total % number_of_slots
    
    # Return the hash value
    return hash_value


def main():
    """Test the hashing function"""

    # Generate a hash value from a string hash key
    key = "A5RD"
    slots = 97
    value = hash_string(key, slots)
    print(f"Hash value for {key} is {value}")


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()
